### PR TITLE
Detail validation errors for loaded-from-file certificate chains

### PR DESCRIPTION
**Pull request by ghulands**
This just adds logging of the openssl failure reason for
checking the certificate chain.

### CODE REVIEW

**Comment by yadij** [comment]
Is there any possible reason than  "ca is not the Issuer for latestCert " ?

This non-success (not error) is ignored because at least one such cert is expected to exist in any PEM files with chains. That being the leaf cert itself.

File: src/security/KeyData.cc [hunk 1]
```diff
@@ -119,14 +119,15 @@ Security::KeyData::loadX509ChainFromFile()
             }
 #endif
             // checks that the chained certs are actually part of a chain for validating cert
-            if (X509_check_issued(ca, latestCert.get()) == X509_V_OK) {
+            int checkCode = X509_check_issued(ca, latestCert.get());
```
Comment (rousskov) [inline src/security/KeyData.cc]:
```suggestion
            const auto checkCode = X509_check_issued(ca, latestCert.get());
```

File: src/security/KeyData.cc [hunk 2]
```diff
@@ -119,14 +119,15 @@ Security::KeyData::loadX509ChainFromFile()
...
+            if (checkCode == X509_V_OK) {
                 debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
                 // OpenSSL API requires that we order certificates such that the
                 // chain can be appended directly into the on-wire traffic.
                 latestCert = CertPointer(ca);
                 chain.emplace_front(latestCert);
             } else {
-                debugs(83, DBG_PARSE_NOTE(2), "Ignoring non-issuer CA from " << certFile << ": " << nameStr);
+                debugs(83, DBG_PARSE_NOTE(2), "Ignoring non-issuer CA from " << certFile << ": " << nameStr << ". Reason: " << checkCode);
```
Comment (rousskov) [inline src/security/KeyData.cc]: If possible, please use X509_verify_cert_error_string() to report the human-friendly version of the numerical error code. If X509_verify_cert_error_string() output does not return the numerical value itself, then please report both.

**Comment by rousskov** [comment]
> Is there any possible reason than "ca is not the Issuer for latestCert " ?

X509_check_issued() failure reasons include: X509_V_ERR_SUBJECT_ISSUER_MISMATCH, X509_V_ERR_KEYUSAGE_NO_DIGITAL_SIGNATURE, X509_V_ERR_KEYUSAGE_NO_CERTSIGN, X509_V_ERR_AKID_SKID_MISMATCH, and X509_V_ERR_AKID_ISSUER_SERIAL_MISMATCH.

Please note that this PR is not changing which errors Squid ignores or reports. This PR details the error. Reporting more error details should be welcomed IMO. If you want to enhance error handling in other ways (e.g., skipping reporting of some specific _expected_ failures), then I suggest doing that in a different PR.

**Comment by rousskov** [comment]
Thank you for addressing my concerns. The next step is for you to rebase this fix on top of master (unless @yadij makes an exception for you). Normally, changes go into master _before_ they land in v4 or other numbered/stable branches. You can read about the overall procedure [eslewhere](https://wiki.squid-cache.org/MergeProcedure).

**Comment by ghulands** [comment]
I have one based off master located here https://github.com/ghulands/squid/tree/tls_debugging. Thanks for the link on the merging procedure - i should have looked for that first. Let me know if you want me to close this PR and open a new one with the change off master or leave both open.

**Comment by rousskov** [comment]
> Let me know if you want me to close this PR and open a new one with the change off master or leave both open.

Either way is fine IMO, but I would leave both open. @yadij will close this one (without prejudice) if it is not needed for v4 backporting.

**Pull request closed without merging**